### PR TITLE
feat: allow users to deregister from signed up jobs.

### PR DESCRIPTION
**Pull request by tcinbis**
This PR adds a unsubscribe button next to the subscribe button. To achieve this the job entity got a new field `allow_unsubscribe` which is set to `False` by default (backwards compatible).

Currently, an unsubscribe will delete all assignments from the user given the job. 

Happy to receive some feedback, especially on the HTML/JS part.

### CODE REVIEW

**Comment by tcinbis** [comment]
@tcinbis fix tests

**Comment by pascalfree** [comment]
Todo

- [ ] Ask for justification when deregistering
- [ ] Set deadline
- [ ] Option to enable by default.

**Comment by tcinbis** [comment]
After our meetup on the 11th of June we decided that we want to add the above mentioned features for de-registering.

I will also check if converting the job form (currently in HTML only) to Django's builtin `Form` class will have significant improvements (thinking of out of the box rendering using `crispy_forms`).

**Comment by pascalfree** [comment]
The failing tests have now been fixed on [release_1.6](https://github.com/juntagrico/juntagrico/tree/release_1.6).
Please merge the branch.

**Comment by siancu** [comment]
What is the status of this PR? Can it be reviewed now?

**Comment by siancu** [comment]
So I had a quick look at this and I have a possibly very stupid question: 

In `jobs.py` I see that a new field is added to the job model object with the default value False. However I cannot see anywhere that this flag is set to True for a specific job? Shouldn't it be set to True when the user subscribes to a job, so that, the next time that the user clicks on the job, the button shows unsubscribe? Or am I missing something here? 

I tried the functionality in the app and it doesn't seem to be working, i.e. I've subscribed to a job, but the status doesn't change and I can even subscribe again and again.

File: juntagrico/views.py [hunk 1]
```diff
@@ -67,27 +67,37 @@ def job(request, job_id):
...
+                elif Config.assignment_unit() == 'HOURS':
+                    amount = job.multiplier * job.duration
+                for i in range(num):
+                    assignment = Assignment.objects.create(
+                        member=member, job=job, amount=amount)
+                for extra in job.type.job_extras_set.all():
+                    if request.POST.get('extra' + str(extra.extra_type.id)) == str(extra.extra_type.id):
+                        assignment.job_extras.add(extra)
+                assignment.save()
+                membernotification.job_signup(member.email, job)
```
Comment (siancu) [inline juntagrico/views.py:89]: I have added here the following two lines:
`job.allow_unsubscribe = True
job.save()`

and it seems to do the trick. Now the unsubscribe button appears next to the subscribe one.

File: juntagrico/views.py [hunk 2]
```diff
@@ -67,27 +67,37 @@ def job(request, job_id):
...
+                # redirect to same page such that refresh in the browser or back
+                # button does not trigger a resubmission of the form
+                return redirect('job', job_id=job_id)
+
+        elif request.POST.get('unsubscribe') and can_unsubscribe:
+            assignment = Assignment.objects.filter(member=member, job=job)
+            assignment.delete()
```
Comment (siancu) [inline juntagrico/views.py:96]: Here I do the opposite:

```
job.allow_unsubscribe = False
job.save()
```

**Comment by tcinbis** [comment]
Hey @siancu, thank you for the comments. 

This PR is still in draft state after the last meet up I attended. We discussed a couple more requirements we would like to add as @pascalfree documented above. 

Regarding your question, the allow_unsubscribe field was supposed to be set by an power user who is creating new jobs and therefore is able to decide whether or not a user should be able to unsubscribe from a job. The reason to set it to false was to not change the expected behavior compared to earlier versions. Does that make sense? :) 

I will check your code suggestions  once I am back home and have access to more than a phone 😃 or feel free to pick up where I left these code changes!

Cheers,
Tom

**Comment by siancu** [comment]
Hi @tcinbis,

Sorry, I missed the other requirements part. You can disregard the code suggestions I made; they were under the impression that, once a user clicked on subscribe, they would themselves be able to unsubscribe from a job, regardless of other things.

**Comment by pascalfree** [comment]
Refactored implementation to use form class.
Also implemented #300 to edit job signup instead of adding to it.
Editing to fewer signups it only allowed if unsubscribe is allowed.
Points in https://github.com/juntagrico/juntagrico/pull/631#issuecomment-1586149956 are still pending.

**Comment by pascalfree** [comment]
> Todo
> 
>     * [ ]  Ask for justification when deregistering
>     * [ ]  Set deadline
>     * [ ]  Option to enable by default.


Rethinking the need for these tasks:
It is actually not the question, if de-registering should be possible or not. It is possible. Currently people write an email to do that, with an apology or explanation.
This PR is easing that process for the admins:

- The slot will automatically become available again, increasing the chance, that someone else will jump in.
- We can control better who will be notified and will not depend on the member sending it to the right email.
- We have a chance to give instructions (in the form), what is expected from the member, when they de-register.

In conclusion: We can soft-release this feature, making it optional. Not per job, but for the entire installation. A deadline for de-registering is not helpful, it just makes the UX inconsistent, so let's not implement that.

New Todos:

* [ ] Ask for justification when de-registering
   * [ ] Notify job contact(s) and all users with a specific notify permission about the de-registration.
* [ ] Display customizable text in the de-registration form
* [ ] Remove option to enable this feature per job
* [ ] Add setting to enable this feature.
